pf2rnaseq/figures/figureHeiserFMSDev.py

```diff
@@ -0,0 +1,186 @@
+
+from ..factorization import pf2
+from ..imports import import_Heiser
+from .common import getSetup, subplotLabel
+
+
+def makeFigure():
+    ax, f = getSetup((6, 3), (1, 3))
+    subplotLabel(ax)
+
+    X = import_Heiser(deviance=True)
+
+    X_lognorm = import_Heiser()
+    percentList = np.arange(0.0, 55.0, 5.0)
+    plot_fms_percent_drop(X, ax[0], percentList=percentList, runs=3)
+
+    ranks = list(range(10, 51, 10))
+    plot_fms_diff_ranks(X, ax[1], ranksList=ranks, runs=3)
+    # plot_fms_diff_ranks_dev_norm(X, X_lognorm, ax[2], ranksList=[5,10,20], runs=2)
+
+    return f
+
+
+def calculateFMS(A: anndata.AnnData, B: anndata.AnnData):
+    """Calculates FMS between 2 factors"""
+    factors = [A.uns["Pf2_A"], A.uns["Pf2_B"], A.varm["Pf2_C"]]
+    A_CP = CPTensor(
+        (
+            A.uns["Pf2_weights"],
+            factors,
+        )
+    )
+
+    factors = [B.uns["Pf2_A"], B.uns["Pf2_B"], B.varm["Pf2_C"]]
+    B_CP = CPTensor(
+        (
+            B.uns["Pf2_weights"],
+            factors,
+        )
+    )
+
+    return fms(A_CP, B_CP, consider_weights=False, skip_mode=1)  # type: ignore
+
+
+def plot_fms_percent_drop(
+    X: anndata.AnnData,
+    ax: Axes,
+    percentList: np.ndarray,
+    runs: int,
+    rank: int = 30,
+):
+    # Plots FMS score when percentage is removed from data
+    dataX = pf2(X, rank, doEmbedding=False)
+
+    fmsLists = []
+
+    for j in range(0, runs, 1):
+        scores = [1.0]
+
+        for i in percentList[1:]:
+            sampled_data: anndata.AnnData = sc.pp.subsample(
+                X, fraction=1 - (i / 100), random_state=j, copy=True
+            )  # type: ignore
+            sampledX = pf2(sampled_data, rank, random_state=j + 2, doEmbedding=False)
+
+            fmsScore = calculateFMS(dataX, sampledX)
+            scores.append(fmsScore)
+
+        fmsLists.append(scores)
+
+    runsList_df = []
+    for i in range(0, runs):
+        for j in range(0, len(percentList)):
+            runsList_df.append(i)
+    percentList_df = []
+    for i in range(0, runs):
+        for j in range(0, len(percentList)):
+            percentList_df.append(percentList[j])
+    fmsList_df = []
+    for sublist in fmsLists:
+        fmsList_df += sublist
+    df = pd.DataFrame(
+        {
+            "Run": runsList_df,
+            "Percentage of Data Dropped": percentList_df,
+            "FMS": fmsList_df,
+        }
+    )
+
+    sns.lineplot(data=df, x="Percentage of Data Dropped", y="FMS", ax=ax)
+    ax.set_ylim(0, 1)
+
+
+def resample(data: anndata.AnnData) -> anndata.AnnData:
+    """Bootstrapping dataset"""
+    indices = np.random.randint(0, data.shape[0], size=(data.shape[0],))
+    data = data[indices].copy()
+    return data
+
+
+def plot_fms_diff_ranks(
+    X: anndata.AnnData,
+    ax: Axes,
+    ranksList: list[int],
+    runs: int,
+):
+    # Plots FMS when using different Pf2 components
+    fmsLists = []
+
+    for j in range(0, runs, 1):
```

Comment:
Seems like this should go in a non-figure file. Also, separate calculation from plotting.